### PR TITLE
refactor: remove user session singleton (part 5) - WPB-4920

### PR DESCRIPTION
**Pull request by agisilaos**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-4920" title="WPB-4920" target="_blank">BugWPB-4920</a>  [iOS] Crash when sending message in C1
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

## Issues

Often the app crashes shortly after logging out or the session expiring.

## Causes

The UI code depends heavily on the active user session, which is accessed via a singleton ZMUserSession.shared(). This is often safely unwrapped, but there are places where it is not, leading to a a crash. The most common point of crash is when the the static function ZMUser.selfUser() is accessed, which depends on the shared user session and implicitly unwraps its value.

In theory, the UI code should be running when there is no user session, but this can not be guaranteed, because the UI code may not have been torn down yet, or part of it may be retained and fired as a block in response to a notification or other observation mechanism.

## Solutions

Avoid all static access to the shared user session and self user. Instead, inject a non optional value from the top level and propagate it where it needs to go via dependency injection.

## Note

Any refinement to better model the user session and dependency structure isn't considered here. First we want to remove the static accessors and make it work. After we can polish it.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
 
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios/Wire-iOS/Sources/UserInterface/CustomAppLock/Unlock/UnlockViewController.swift [hunk 1]
```diff
@@ -32,7 +32,7 @@ protocol UnlockViewControllerDelegate: AnyObject {
 /// This VC should be wrapped in KeyboardAvoidingViewController as the "unlock" button would be covered on 4 inch iPhone
 final class UnlockViewController: UIViewController {
```
Comment (johnxnguyen) [inline wire-ios/Wire-iOS/Sources/UserInterface/CustomAppLock/Unlock/UnlockViewController.swift:34]: It'd be better if we only use `UserSession` and not the other protocols.

File: wire-ios/Wire-iOS/Sources/UserInterface/StartUI/StartUI/StartUIViewController.swift [hunk 2]
```diff
@@ -207,7 +207,7 @@ final class StartUIViewController: UIViewController, SpinnerCapable {
 
     func showKeyboardIfNeeded() {
         // TODO: unwrap
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/StartUI/StartUI/StartUIViewController.swift]: remove todo

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientListViewController.swift [hunk 3]
```diff
@@ -128,7 +130,11 @@ final class UserClientListViewController: UIViewController, UICollectionViewDele
 
     func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
         let client = clients[indexPath.row]
-        let profileClientViewController = ProfileClientViewController(client: client as! UserClient, fromConversation: true) // TODO jacob don't force unwrap
+        let profileClientViewController = ProfileClientViewController(
+            client: client as! UserClient,
+            fromConversation: true, 
+            userSession: userSession
+        ) // TODO jacob don't force unwrap
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientListViewController.swift]:
``` suggestion
        guard let client = clients[indexPath.row] as? UserClient else { return }
        let profileClientViewController = ProfileClientViewController(
            client: client,
            fromConversation: true, 
            userSession: userSession
        )
```

**Review by netbe — APPROVED** [review]
just some cleanup todos

File: wire-ios/Wire-iOS/Sources/UserInterface/CustomAppLock/Unlock/UnlockViewController.swift (hunk 1 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/CustomAppLock/Unlock/UnlockViewController.swift:34]: I've done that in part 6. So for now we're good! @johnxnguyen

File: wire-ios/Wire-iOS/Sources/UserInterface/StartUI/StartUI/StartUIViewController.swift (hunk 2 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/StartUI/StartUI/StartUIViewController.swift]: Done!

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientListViewController.swift (hunk 3 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientListViewController.swift]: Done!

**Review by johnxnguyen — APPROVED** [review]

**Pull request merged**